### PR TITLE
WIP: Yellow Vault Upgradability Contracts

### PR DESCRIPTION
**Pull request by nksazonov**

### CODE REVIEW

File: contracts/yellow/VaultImplBase.sol [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    return __newerImplementation;
+  }
+
+  function setNewerImplementation(address newerImplementation) external notDelegated onlyAdmin {
+    require(__newerImplementation == address(0), 'newerImplementation is already set');
+    // prevent unnecessary event emitions & broken newerImplementaion chain
+    require(
+      newerImplementation != __newerImplementation && newerImplementation != __self,
+      'invalid newerImplementation supplied'
+    );
```
Comment (lckmidnight) [inline contracts/yellow/VaultImplBase.sol:61]: Change `newerImplementation != __newerImplementation` check to `newerImplementation != address(0)`

If `__newerImplementation` is not zero, this mean that the first `require` check will revert, so we only need to check if `newerImplementation` parameter is not `address(0)` or `__self`

**Review by lckmidnight — CHANGES_REQUESTED** [review]

File: test/yellow/VaultUpgradability.spec.ts [hunk 2]
```diff
@@ -0,0 +1,456 @@
...
+    });
+
+    it('revert on setting newer implementation to same address', async () => {
+      expect(await VaultImpl1.getNewerImplementation()).to.be.equal(AddressZero);
+      const newerImplAddress = Wallet.createRandom().address;
+      await VaultImpl1.connect(implAdmin).setNewerImplementation(newerImplAddress);
+      await expect(
+        VaultImpl1.connect(implAdmin).setNewerImplementation(newerImplAddress)
+      ).to.be.revertedWith(NEWER_IMPL_IS_SET);
+    });
```
[the remark this answers is not in the document]
Comment (nksazonov) [reply test/yellow/VaultUpgradability.spec.ts]: bc323135

File: test/yellow/VaultUpgradability.spec.ts [hunk 3]
```diff
@@ -0,0 +1,456 @@
...
+      await expect(VaultImpl1Proxied.connect(user).getNewerImplementation()).to.be.revertedWith(
+        MUST_NOT_THROUGH_DELEGATECALL
+      );
+    });
+
+    it('revert on call `setNewerImplementation` via proxy', async () => {
+      await expect(
+        VaultImpl1Proxied.connect(implAdmin).setNewerImplementation(VaultImpl2.address)
+      ).to.be.revertedWith(MUST_NOT_THROUGH_DELEGATECALL);
+    });
```
[the remark this answers is not in the document]
Comment (nksazonov) [reply test/yellow/VaultUpgradability.spec.ts]: 4b95738

File: contracts/yellow/test/TESTVaultProxy.sol [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+import '@openzeppelin/contracts/access/AccessControl.sol';
+import '@openzeppelin/contracts/proxy/Proxy.sol';
+import '@openzeppelin/contracts/proxy/ERC1967/ERC1967Upgrade.sol';
+import '../VaultImplBase.sol';
+
+/**
+ * @dev Use for TEST PURPOSES ONLY. !!! Contains security vulnerability !!!
+ */
+contract TESTVaultProxy is Proxy, ERC1967Upgrade { // Admin AccessControl logic is present in ERC1967Proxy itself
```
[the remark this answers is not in the document]
Comment (nksazonov) [reply contracts/yellow/test/TESTVaultProxy.sol]: Just didn't have enough time to did it yesterday 😄 
301bb7b

File: contracts/yellow/VaultImplBase.sol (hunk 1 above)
Comment (nksazonov) [reply contracts/yellow/VaultImplBase.sol:61]: Nice catch!

Comment (nksazonov) [reply contracts/yellow/VaultImplBase.sol]: f409316

File: contracts/yellow/VaultImplBase.sol [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    emit NewerImplementationSet(newerImplementation);
+  }
+
+  function getAdmin() external view notDelegated returns (address) {
+    return ERC1967Upgrade._getAdmin();
+  }
+
+  function changeAdmin(address newAdmin) external onlyAdmin notDelegated {
+    ERC1967Upgrade._changeAdmin(newAdmin);
+  }
```
[the remark this answers is not in the document]
Comment (nksazonov) [reply contracts/yellow/VaultImplBase.sol]: 47f61c36

File: contracts/yellow/VaultImplBase.sol [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+    emit NewerImplementationSet(newerImplementation);
+  }
+
+  function getAdmin() external view returns (address) {
+    return ERC1967Upgrade._getAdmin();
+  }
+
+  function changeAdmin(address newAdmin) external onlyAdmin {
+    ERC1967Upgrade._changeAdmin(newAdmin);
+  }
```
Comment (lckmidnight) [inline contracts/yellow/VaultImplBase.sol:74]: Can you move this above `// Implementation context storage`

In a specific `// Shared context` or something like this, to emphasis that this logic is shared for both Impl and Proxy contexts

File: contracts/yellow/VaultProxyBase.sol [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+      newImplementation,
+      abi.encodeWithSelector(bytes4(keccak256("initialize()"))),
+      true
+    );
+  }
+
+  modifier onlyAdmin() {
+    require(msg.sender == ERC1967Upgrade._getAdmin(), 'caller not admin');
+    _;
+  }
```
Comment (lckmidnight) [inline contracts/yellow/VaultProxyBase.sol:40]: This can be removed

File: test/yellow/VaultUpgradability.spec.ts [hunk 8]
```diff
@@ -0,0 +1,489 @@
...
+
+    it('revert on setting the same newer implementation twice', async () => {
+      expect(await VaultImpl1.getNewerImplementation()).to.be.equal(AddressZero);
+      const newerImplAddress = Wallet.createRandom().address;
+      await VaultImpl1.connect(implAdmin).setNewerImplementation(newerImplAddress);
+      expect(await VaultImpl1.getNewerImplementation()).to.be.equal(newerImplAddress);
+      await expect(
+        VaultImpl1.connect(implAdmin).setNewerImplementation(newerImplAddress)
+      ).to.be.revertedWith(NEWER_IMPL_IS_SET);
+    });
```
Comment (lckmidnight) [inline test/yellow/VaultUpgradability.spec.ts:113]: Again, these tests are duplicates,

the test case here should be to be sure that an address can not be set twice, we don't care if it's the same or a different one.

**Review by lckmidnight — CHANGES_REQUESTED** [review]

File: contracts/yellow/VaultImplBase.sol (hunk 6 above)
Comment (nksazonov) [reply contracts/yellow/VaultImplBase.sol:74]: Will address this in dev docs commit

File: contracts/yellow/VaultProxyBase.sol (hunk 7 above)
Comment (nksazonov) [reply contracts/yellow/VaultProxyBase.sol]: 6796a0f7

File: contracts/yellow/VaultImplBase.sol (hunk 6 above)
Comment (nksazonov) [reply contracts/yellow/VaultImplBase.sol]: 7b2e745

**Pull request merged**